schachmat/wego
# feat: Add caching mechanism to optimize weather data fetching from Op…

Issue: #196 

Motivation and Context

This PR introduces caching to the weather backend using the OpenMeteo API. It improves performance by storing weather data for each location, reducing the number of API calls and ensuring data freshness for a specific period. It addresses the need to limit repeated API requests within a short time frame, enhancing the application's efficiency.

Key Benefits:

Reduced API request load.
Faster response times for repeated requests.
Improved resource management and performance.
Description

Added caching functionality:

Introduced a cache map to store weather data by location, along with timestamps.
Data remains valid for a specified cacheDuration (set to 1 hour).
Before making API calls, the cache is checked for existing, valid data.
Enhanced the Fetch function:

It now checks the cache before fetching fresh data from the OpenMeteo API.
If the cache contains valid data (based on timestamp), it returns cached data.
Otherwise, it fetches new data, updates the cache, and returns the updated result.
Synchronization using mutex locks:

Added a sync.Mutex to ensure thread-safe access to the cache map, preventing race conditions in concurrent requests.
Steps for Testing

Setup:

Pull the latest changes from this branch.
Ensure the Go environment is properly set up with required dependencies (e.g., github.com/schachmat/wego/iface).
Testing the caching functionality:

Run the backend and make a request to the OpenMeteo API for a specific location.
Check the logs to confirm that a fresh API request is made.
Wait less than 1 hour and make the same request for the same location again. The cache should be used, reducing the API call.
Wait for more than 1 hour, and ensure that a new API request is made after cache expiration.
Check concurrency:

Simulate multiple concurrent requests for the same location to verify that caching works without race conditions or deadlocks.
Screenshots (if applicable)

N/A for this PR (backend changes are not visual).
